backend/src/schema/resolvers/posts.spec.js

```diff
@@ -383,3 +400,315 @@ describe('DeletePost', () => {
+        await postMutationAction(user, addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          postQueryAction(PostsEmotionsCountQuery, postsEmotionsQueryVariables),
+        ).resolves.toEqual(expect.objectContaining(expected))
+      })
+
+      it('allows a user to add more than one emotion', async () => {
+        const expectedEmotions = [
+          { emotion: 'happy', User: { id: user.id } },
+          { emotion: 'surprised', User: { id: user.id } },
+        ]
+        const expectedResponse = {
+          data: { Post: [{ emotions: expect.arrayContaining(expectedEmotions) }] },
+        }
+        await postMutationAction(user, addPostEmotionsMutation, addPostEmotionsVariables)
+        addPostEmotionsVariables.data.emotion = 'surprised'
+        await postMutationAction(user, addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          postQueryAction(PostsEmotionsQuery, postsEmotionsQueryVariables),
+        ).resolves.toEqual(expect.objectContaining(expectedResponse))
+      })
+    })
+
+    describe('authenticated as author', () => {
+      beforeEach(() => {
+        user = owner
+      })
+
+      it('adds an emotion to the post', async () => {
+        const expected = {
+          data: {
+            AddPostEmotions: {
+              from: { id: owner.id },
+              to: addPostEmotionsVariables.to,
+              emotion: 'happy',
+            },
+          },
+        }
+        await expect(
+          postMutationAction(user, addPostEmotionsMutation, addPostEmotionsVariables),
+        ).resolves.toEqual(expect.objectContaining(expected))
+      })
+    })
+  })
+
+  describe('RemovePostEmotions', () => {
+    let removePostEmotionsVariables, postsEmotionsQueryVariables
+    const removePostEmotionsMutation = gql`
+      mutation($to: _PostInput!, $data: _EMOTEDInput!) {
+        RemovePostEmotions(to: $to, data: $data) {
+          from {
+            id
+          }
+          to {
+            id
+          }
+          emotion
+        }
+      }
+    `
+    beforeEach(async () => {
+      await ownerNode.relateTo(postToEmoteNode, 'emoted', { emotion: 'cry' })
+      await postMutationAction(user, addPostEmotionsMutation, addPostEmotionsVariables)
+
+      postsEmotionsQueryVariables = { id: postToEmote.id }
+      removePostEmotionsVariables = {
+        to: { id: postToEmote.id },
+        data: { emotion: 'cry' },
+      }
+    })
+
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        user = null
+        const removePostEmotions = await postMutationAction(
+          user,
+          removePostEmotionsMutation,
+          removePostEmotionsVariables,
+        )
+        expect(removePostEmotions.errors[0]).toHaveProperty('message', 'Not Authorised!')
+      })
+    })
+
+    describe('authenticated', () => {
+      describe('but not the emoter', () => {
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-08-05T14:22:04Z" title="Monday, August 5th 2019, 4:22:04 pm +02:00">Aug 5, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -383,3 +385,220 @@ describe('DeletePost', () => {
     })
   })
 })
+
+describe('emotions', () => {
+  let addPostEmotionsVariables
+  let postEmotionsQueryVariables
+  const postEmotionsQuery = `
+    query($id: ID!) {
+      Post(id: $id) {
+        emotions {
+          emotion
+          User {
+            id
+          }
+        }
+      }
+    }
+  `
+  beforeEach(async () => {
+    const asAuthor = Factory()
+    authorParams = {
+      id: authorParams.id,
+      email: 'wanna-add-emotions@example.org',
+      password: '1234',
+    }
+    await asAuthor.create('User', authorParams)
+    await asAuthor.authenticateAs(authorParams)
+    await asAuthor.create('Post', {
+      id: 'p1376',
+      title: postTitle,
+      content: postContent,
+    })
+    postEmotionsQueryVariables = { id: 'p1376' }
+  })
+
+  describe('AddPostEmotions', () => {
+    beforeEach(() => {
+      addPostEmotionsVariables = {
+        from: { id: authorParams.id },
+        to: { id: 'p1376' },
+        data: { emotion: 'happy' },
+      }
+    })
+
+    const addPostEmotionsMutation = `
+      mutation($from: _UserInput!, $to: _PostInput!, $data: _EMOTEDInput!) {
+        AddPostEmotions(from: $from, to: $to, data: $data) {
+          from { id }
+          to { id }
+          emotion
+        }
+      }
+    `
+    const postEmotionsCountQuery = `
+      query($id: ID!) {
+        Post(id: $id) {
+          emotionsCount
+        }
+      }
+    `
+
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        client = new GraphQLClient(host)
+        await expect(
+          client.request(addPostEmotionsMutation, {
+            from: { id: authorParams.id },
+            to: { id: 'p1376' },
+            data: { emotion: 'happy' },
+          }),
+        ).rejects.toThrow('Not Authorised')
+      })
+    })
+
+    describe('authenticated and not the author', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(userParams)
+        client = new GraphQLClient(host, { headers })
+      })
+
+      it('adds an emotion to the post', async () => {
+        addPostEmotionsVariables.from.id = userParams.id
+        const expected = {
+          AddPostEmotions: {
+            from: addPostEmotionsVariables.from,
+            to: addPostEmotionsVariables.to,
+            emotion: 'happy',
+          },
+        }
+        await expect(
+          client.request(addPostEmotionsMutation, addPostEmotionsVariables),
+        ).resolves.toEqual(expected)
+      })
+
+      it('limits the addition of the same emotion to 1', async () => {
+        const expected = {
+          Post: [
+            {
+              emotionsCount: 1,
+            },
+          ],
+        }
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          client.request(postEmotionsCountQuery, postEmotionsQueryVariables),
+        ).resolves.toEqual(expected)
+      })
+
+      it('allows a user to add more than one emotion', async () => {
+        const expected = [
+          { emotion: 'happy', User: { id: authorParams.id } },
+          { emotion: 'surprised', User: { id: authorParams.id } },
+        ]
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        addPostEmotionsVariables.data.emotion = 'surprised'
+        await client.request(addPostEmotionsMutation, addPostEmotionsVariables)
+        await expect(
+          client.request(postEmotionsQuery, postEmotionsQueryVariables),
+        ).resolves.toEqual({
+          Post: [{ emotions: expect.arrayContaining(expected) }],
+        })
+      })
+    })
+
+    describe('authenticated as author', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(authorParams)
+        client = new GraphQLClient(host, { headers })
+      })
+
+      it('adds an emotion to the post', async () => {
+        const expected = {
+          AddPostEmotions: {
+            from: addPostEmotionsVariables.from,
+            to: addPostEmotionsVariables.to,
+            emotion: 'happy',
+          },
+        }
+        await expect(
+          client.request(addPostEmotionsMutation, addPostEmotionsVariables),
+        ).resolves.toEqual(expected)
+      })
+    })
+  })
+
+  describe('RemovePostEmotions', () => {
+    let removePostEmotionsVariables
+    const removePostEmotionsMutation = `
+      mutation($from: _UserInput!, $to: _PostInput!, $data: _EMOTEDInput!) {
+        RemovePostEmotions(from: $from, to: $to, data: $data)
+      }
+    `
+    beforeEach(() => {
+      removePostEmotionsVariables = {
+        from: { id: authorParams.id },
+        to: { id: 'p1376' },
+        data: { emotion: 'cry' },
+      }
+    })
+
+    describe('unauthenticated', () => {
+      it('throws authorization error', async () => {
+        client = new GraphQLClient(host)
+        await expect(
+          client.request(removePostEmotionsMutation, removePostEmotionsVariables),
+        ).rejects.toThrow('Not Authorised')
+      })
+    })
+
+    describe('authenticated', () => {
+      let headers
+      beforeEach(async () => {
+        headers = await login(authorParams)
+        client = new GraphQLClient(host, { headers })
+        await factory.emote({
+          from: authorParams.id,
+          to: 'p1376',
+          data: 'cry',
+        })
+        await factory.emote({
+          from: authorParams.id,
+          to: 'p1376',
+          data: 'happy',
+        })
+      })
+
+      describe('but not the emoter', () => {
```


I love the word `emoter `! 😍